gamemodes/terrortown/gamemode/server/sv_player.lua

```diff
@@ -511,88 +513,89 @@ function GM:KeyPress(ply, key)
-    local tr = util.TraceLine({
-        start = ply:GetShootPos(),
-        endpos = ply:GetShootPos() + ply:GetAimVector() * 100,
-        filter = ply,
-        mask = MASK_SHOT,
-    })
+local function SpecUseKey(ply, ent)
+    -- Do not allow the spectator to gather information if they're about to revive.
+    if ply:IsReviving() then
+        LANG.Msg(ply, "spec_about_to_revive", nil, MSG_MSTACK_WARN)
 
-    if not tr.Hit or not IsValid(tr.Entity) then
         return
     end
 
-    if tr.Entity.CanUseKey and tr.Entity.UseOverride then
-        local phys = tr.Entity:GetPhysicsObject()
-
-        if IsValid(phys) and not phys:HasGameFlag(FVPHYSICS_PLAYER_HELD) then
-            tr.Entity:UseOverride(ply)
-
-            return true
+    if ent.player_ragdoll then
+        if not ply:KeyDown(IN_WALK) then
+            CORPSE.ShowSearch(ply, ent)
         else
-            -- do nothing, can't +use held objects
-            return true
+            ply:Spectate(OBS_MODE_IN_EYE)
+            ply:SpectateEntity(ent)
         end
-    elseif tr.Entity.player_ragdoll then
-        CORPSE.ShowSearch(ply, tr.Entity, ply:KeyDown(IN_WALK) or ply:KeyDownLast(IN_WALK)) -- Body Corpse Search Identify TODO
-
-        return true
+    elseif ent:IsPlayer() and ent:IsActive() then
+        ply:Spectate(ply.spec_mode or OBS_MODE_IN_EYE)
+        ply:SpectateEntity(ent)
+    else
+        PROPSPEC.Target(ply, ent)
     end
 end
 
 ---
--- Normally all dead players are blocked from IN_USE on the server, meaning we
--- can't let them search bodies. This sucks because searching bodies is
--- fun. Hence on the client we override +use for specs and use this instead.
-local function SpecUseKey(ply, cmd, arg)
-    if not IsValid(ply) or not ply:IsSpec() then
+-- This is called by a client when using the "+use"-key
+-- and contains the entity which was detected
+-- @param number len Length of the message
+-- @param Player ply Player that sent the message
+-- @realm server
+-- @internal
+net.Receive("TTT2PlayerUseEntity", function(len, ply)
+    local ent = net.ReadEntity()
+
+    if not IsValid(ent) then
         return
     end
 
-    -- Do not allow the spectator to gather information if they're about to revive.
-    if ply:IsReviving() then
-        LANG.Msg(ply, "spec_about_to_revive", nil, MSG_MSTACK_WARN)
-
+    -- Check if the use interaction is possible
+    -- Use 150 instead of 100 to allow some desync between client and server
+    local distance = ply:GetShootPos():Distance(ent:GetPos())
+    if distance > 150 then
         return
     end
 
-    -- longer range than normal use
-    local tr = util.QuickTrace(ply:GetShootPos(), ply:GetAimVector() * 128, ply)
+    ---
+    -- Enable addons to allow handling PlayerUse
+    -- Otherwise default to old IsTerror Check
+    -- @realm server
+    -- stylua: ignore
+    if hook.Run("PlayerUse", ply, ent, ply:IsTerror()) then
+        ent:Use(ply, ply)
+    end
+    overrideDoPlayerUse = false
```

Comment:
I think that line causes a bug. Because it now declared a global name with the same name as the parameter in the hook. Meaning that the UseHook will now probably always return false when not called by the hook.Run above.

But I think this is only a leftover line anyway?

Reply:
Oops, good catch. Could have been a nasty bug to find. Not sure what would happen if the global variable is set and the parameter is nil.